src/java/test/org/apache/zookeeper/server/quorum/SessionUpgradeTest.java

```diff
@@ -0,0 +1,516 @@
+                localSessionId, localSessionPwd);
+        watcher.waitForConnected(CONNECTION_TIMEOUT);
+
+        // The created ephemeral nodes are still around.
+        for (int i = 0; i < 5; i++) {
+            Assert.assertNotNull(zk.exists(nodePrefix + i, null));
+        }
+
+        // When we explicitly close the session, we should not be able to
+        // reconnect with the same session id
+        zk.close();
+
+        try {
+            watcher.reset();
+            zk = new DisconnectableZooKeeper(
+                    hostPorts[otherFollowerIdx], CONNECTION_TIMEOUT, watcher,
+                    localSessionId, localSessionPwd);
+            zk.exists(nodePrefix + "0", null);
+            Assert.fail("Reconnecting to a closed session ID should fail.");
+        } catch (KeeperException.SessionExpiredException e) {
+        }
+
+        watcher.reset();
+        // And the ephemeral nodes will be gone since the session died.
+        zk = new DisconnectableZooKeeper(
+                hostPorts[testPeerIdx], CONNECTION_TIMEOUT, watcher);
+        watcher.waitForConnected(CONNECTION_TIMEOUT);
+        for (int i = 0; i < 5; i++) {
+            Assert.assertNull(zk.exists(nodePrefix + i, null));
+        }
+    }
+
+    @Test
+    public void testLocalSessionUpgradeSnapshot() throws IOException, InterruptedException {
+        // setup the env with RetainDB and local session upgrading
+        ClientBase.setupTestEnv();
+
+        final int SERVER_COUNT = 3;
+        final int clientPorts[] = new int[SERVER_COUNT];
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            clientPorts[i] = PortAssignment.unique();
+            sb.append("server.").append(i).append("=127.0.0.1:")
+              .append(PortAssignment.unique()).append(":")
+              .append(PortAssignment.unique()).append("\n");
+        }
+        sb.append("localSessionsEnabled=true\n");
+        sb.append("localSessionsUpgradingEnabled=true\n");
+        String cfg = sb.toString();
+
+        // create a 3 server ensemble
+        MainThread mt[] = new MainThread[SERVER_COUNT];
+        final TestQPMainDropSessionUpgrading qpMain[] =
+                new TestQPMainDropSessionUpgrading[SERVER_COUNT];
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            final TestQPMainDropSessionUpgrading qp = new TestQPMainDropSessionUpgrading();
+            qpMain[i] = qp;
+            mt[i] = new MainThread(i, clientPorts[i], cfg, false) {
+                @Override
+                public TestQPMain getTestQPMain() {
+                    return qp;
+                }
+            };
+            mt[i].start();
+        }
+
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            Assert.assertTrue("waiting for server " + i + " being up",
+                    ClientBase.waitForServerUp("127.0.0.1:" + clientPorts[i],
+                            CONNECTION_TIMEOUT));
+        }
```

Comment:
Test initialization logic is better placed in setUp() method especially when it's common in multiple tests.
(That would be one benefit of moving new tests to new file)